lib/ansible/modules/windows/win_domain_controller.ps1

```diff
@@ -208,7 +208,30 @@ Try {
                 if ($site_name) {
                     $install_params.SiteName = $site_name
                 }
-                $install_result = Install-ADDSDomainController -NoRebootOnCompletion -Force @install_params
+                try
+                {
+                    $install_result = Install-ADDSDomainController -NoRebootOnCompletion -Force @install_params
+                } catch [Microsoft.DirectoryServices.Deployment.DCPromoExecutionException] {
+                    # ExitCode 15 == 'Role change is in progress or this computer needs to be restarted.'
+                    # DCPromo exit codes details can be found at https://docs.microsoft.com/en-us/windows-server/identity/ad-ds/deploy/troubleshooting-domain-controller-deployment
+                    if ($_.Exception.ExitCode -eq 15) {
+                        $result.reboot_required = $true
+                    } else {
+                        Fail-Json -obj $result -message "Failed to install ADDSDomainController with DCPromo: $($_.Exception.Message)"
+                    }
+                } catch {
+                    # We cannot directly catch Microsoft.ADRoles.Deployment.Common.Tests.TestFailedException
+                    # As the type is within an internal flagged class of Microsoft.ADRoles.Deployment.Common
+                    if (
```

Comment:
I don't see a reason why we need this special handling here, when I test it locally the exception already shows the failed to resolve domain like message

```
Install-ADDSDomainController : Verification of user credential permissions failed. An Active Directory domain
controller for the domain "fake.domain.com" could not be contacted.
Ensure that you supplied the correct DNS domain name.
At line:1 char:1
+ Install-ADDSDomainController -DomainName 'fake.domain.com' -NoRebootO ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Install-ADDSDomainController], TestFailedException
    + FullyQualifiedErrorId : Test.VerifyUserCredentialPermissions.DCPromo.General.25,Microsoft.DirectoryServices.Depl
   oyment.PowerShell.Commands.InstallADDSDomainControllerCommand
```

It does have some extra fluff but I don't think we need to complicate the catch logic especially since this is an internal exception.

Reply:
I just did not want to do a catch all and this was the best solution I found to only match that particular exception. How could this be improved?

Reply:
Honestly just don’t catch all as Ansible will now properly display the traceback and error message for you. You should only catch exceptions if the actual error is vague and you can improve it.

Reply:
But if I don't catch it the playbook fails. And exactly when this error is thrown that should not happen.

Reply:
Oh keep the first exception handler for DCPromoExecutionEngine but just remove the whole generic exception catch block.

Reply:
But than the user gets an error that cannot be handled within the playbook (and the whole play will fail) if Install-ADDSDomainController is unable to resolve the dns domain name... With the current exception handling a user could catch the error within the playbook and take action.

Reply:
You can still pick up unhandled exceptions in a task, we even have a test for it. [This test](https://github.com/ansible/ansible/blob/devel/test/integration/targets/win_exec_wrapper/tasks/main.yml#L28-L39) shows you what happens when you don't catch an explicit throw. The `msg` return value is the string `Unhandled exception while executing module: $($_.Exception.Message)` whereas the `exception` return value is the PowerShell error message and script stack trace.

Then [this test](https://github.com/ansible/ansible/blob/devel/test/integration/targets/win_exec_wrapper/tasks/main.yml#L67-L78) shows you what happens when a .net function raises an unhandled exception. The same thing occurs and you can use `msg` or `exception` to parse.

Here is the module code that it is running to show you how the errors are fired https://github.com/ansible/ansible/blob/devel/test/integration/targets/win_exec_wrapper/library/test_fail.ps1.

So for this case you would have something like

```
- win_domain_controller:
    dns_domain_name: unknown.domain
    domain_admin_user: username@unknown.domain
    domain_admin_password: password123!
    safe_mode_password: password123!
    state: domain_controller
  register: dc_result
  failed_when: not dc_result is successful or dc_result.msg != "could not be contacted"
```

Reply:
I still don't agree to leave out that catch, but that may be something for my internal fork and not something for general use. I'm going to remove it for now.